maintenance/zero_downtime_migration.rst

```diff
@@ -0,0 +1,167 @@
+
+2. For implementations requiring sharding, execute the following configuration steps:
+
+   - Ensure your MongoDB installation supports sharding. Refer to the MongoDB documentation if uncertain.
+   - Provision the schema using the following command:
+
+      .. code-block:: bash
+
+         COLLECTION_NAME=vonkentries
+         CONNECTION_STRING=<YOUR_CONNECTION_STRING>
+         LICENSE_FILE=<path-to-your-license-file>
+
+         mkdir empty_dir
+
+         dotnet fsi.dll \
+         --provisionTargetDatabase true \
+         --dbType MongoDb \
+         --mongoConnectionstring $CONNECTION_STRING \
+         --mongoCollection $COLLECTION_NAME \
+         --license $LICENSE_FILE \
+         --source empty_dir
+
+         rm -rf empty_dir
+
+   - Configure sharding for the entries collection using the command below:
+
+      .. code-block:: bash
+
+         DB_NAME=vonkdata
+         COLLECTION_NAME=vonkentries
+         CONNECTION_STRING=<YOUR_CONNECTION_STRING>
+
+         mongosh $CONNECTION_STRING <<EOF
+         sh.shardCollection("$DB_NAME.$COLLECTION_NAME", { type: 1, im: 1, cur: 1, cnt: 1, change: 1, res_id: "hashed" });
+         EOF
+
+3. Execute the FSI migration process:
+  
+   Initialize the migration by executing the following command:
+
+   .. code-block:: bash
+
+        RECOVERY_JOURNAL_DIRECTORY=./journal # A directory where the progress will be stored in case of a crash. FSI will quickly catch up to the place where an error occurred.
+        FHIR_VERSION=R4
+
+        LICENSE_FILE="<path to your license file>"
+
+        RUNNING_MODE=Continuous # Or AdHoc. If AdHoc mode is used, FSI will terminate when all the items from the old DB have been processed.
+
+        SOURCE_CONNECTION_STRING="<old db connection string>"
+        SOURCE_COLLECTION_NAME=vonkentries
+
+        CONNECTION_STRING="<new DB connection string>"
+        COLLECTION_NAME=vonkentries
+
+        dotnet fsi.dll \
+            --fhir-version $FHIR_VERSION \
+            --provisionTargetDatabase true \
+            --useRecoveryJournal $RECOVERY_JOURNAL_DIRECTORY \
+            --sourceType MongoDb \
+            --srcMongoCollection $SOURCE_COLLECTION_NAME \
+            --srcMongoConnectionString $SOURCE_CONNECTION_STRING \
+            --srcMongoRunningMode $RUNNING_MODE \
+            --update-existing-resources ErrorOnConflict \
+            --dbType MongoDb \
+            --mongoConnectionstring $CONNECTION_STRING \
+            --mongoCollection $COLLECTION_NAME \
+            --license $LICENSE_FILE
+
+   Migration completion is indicated by the following message: ``No new items found in the database. Waiting for 00:00:05 before retrying...``
+
+4. Provision the new version of Firely Server instance
+5. Update the reverse proxy configuration to direct traffic to the new Firely Server instance
+6. Upon successful migration verification, terminate the FSI migration tool and decommission the previous Firely Server installation.
```

Comment:
I would make this a bit more elaborate:
4. Provision an instance of the new version of Firely Server
5. Verify that the migration was successful by inspecting the data in the target database, see the Verification Steps below.
6. Redirect traffic...
7. Decommission the instance(s) of the old version of Firely Server
8. Terminate the FSI migration tool

Note that I also reversed the order of steps 7 and 8 to make sure nothing new comes in through the old FS after FSI was terminated. Although the routing by the proxy should also avoid that of course.

Reply:
Numbering was off, but I fixed that for you.